Various Patches (L10N boss num, selList overflow, chat middle mention+refocus on click)

Helcostr opened the pull request:
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
* Fixes #9470
* Fixes #12891
* Fixes #12377

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
* Uses browser JS `toLocaleString` to convert numbers to appropriately "marked" numbers
* Limit `membersModal` list selector to 12-unit grid (to prevent wrap around on long text for `sortOptions`)
* Add ellipse based overflow for `membersModal` list selector
* Add padding so that the checkmark for `membersModal` list doesn't overlap with text
* Allow adding `@mention` anywhere (previously was fixed at the end of the text)
* Add re-focus after mouseclick on `@mention` choice



[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 10cbe941-0e10-4763-9f4f-7a3ae5bb2198

Helcostr committed:
L10N boss numbers

Helcostr committed:
Limit to 12-unit grid, test selList CSS (not working)

Helcostr committed:
Fix selList overflow to use ellipsis

Helcostr committed:
Allow adding @metion in middle of chat.
Add re-focus after mouseclick.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -128,38 +130,29 @@ export default {
-    text (newText) {
-      if (!newText[newText.length - 1] || newText[newText.length - 1] === ' ') {
+    text (newTextParam, prevTextParam) {
+      const delCharFocusBool = prevTextParam.length > newTextParam.length;
```
I would name this something like `wereCharsRemoved` or `isDeleting`

Helcostr replied:
Agreed.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -128,38 +130,29 @@ export default {
+    text (newTextParam, prevTextParam) {
+      const delCharFocusBool = prevTextParam.length > newTextParam.length;
+      const caretPosition = this.textbox.selectionEnd;
+      const charFocus = (
+        delCharFocusBool ? prevTextParam : newTextParam
+      )[caretPosition - (delCharFocusBool ? 0 : 1)];
```
I think this line would be clearer if you stick to a single ternary (And I'd recommend renaming `charFocus` to something like `lastFocussedChar`)

```suggestion
      const lastFocussedChar = isDeleting ? prevText[caretPosition] : newText[caretPosition - 1];
```

Helcostr replied:
Ah yeah, this was relic code that I had when I stored to focusText (including the focusChar)

Which is why the double ternary still exists.

Good suggestion.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -128,38 +130,29 @@ export default {
   watch: {
-    text (newText) {
-      if (!newText[newText.length - 1] || newText[newText.length - 1] === ' ') {
+    text (newTextParam, prevTextParam) {
```
I don't think adding `Param` to parameters is adding useful information to the parameter names.  It's not something I've seen in the rest of the codebase so I'd recommend leaving it out here as well.

Helcostr replied:
It's to prevent conflicting naming schemes, since I noticed a few lines up (or down, I don't remember what direction), newText is a totally different meaning: it means text that the programmer generated, not the new text as part of the parameter of watch.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -90,12 +90,14 @@ export default {
-      atRegex: /(?!\b)@[\w-]*$/,
+      // Old Regex
+      // atRegex: /(?!\b)@([\w-]*)$/,
+      atRegex: /@([\w-]*)$/,
```
Why remove the negative word boundary lookahead?  I agree that a positive lookbehind for negated word boundary would probably make more sense but the search should not start when typing e.g. an e-mail address right?

Ah is this to enable searches anywhere in the text?  If that's the case you should drop the end-of-text anchor, however that worries me that you might find earlier mentions in the same text block with this regex and the selection will jump back to that mention instead.  I didn't run your branch yet, so not sure, but would be good to check.

Also it's not necessary to leave old code around.  It's all in the git history anyway.

Helcostr replied:
Hehe... my secret weapon... XD

First of all, negative lookahead is awkward since the mentions trigger without boundaries, so typing an email is something I gotta let slide. If the user is annoyed while typing an email, they can escape the search (or the search will pop up invalid).

> Try typing `@mention1@mention2` and sending it. It works. The script will auto insert spaces, but if ur a human typing, u can type w/o the spaces. I just did think of an edge case that I need to check though... thx for reminding me. Edit: I think the edge case I was thinking of is fine... can't remember what the issue was exactly anyways...

The end of line is necessary to find the last mention of the textblock, since I didn't start the regex with a "start here" index.

Also old code: whoops, missed cleaning the regex comment when I was keeping it around as a backup (a big flag for me)

The comment issue has been resolved. Looking for feedback on the regex in question.

benkelaar replied:
I'm not sure what you mean, the whole point of the negative lookahead is to make the mentions trigger only without boundaries right?  Historically the choice was made to favour not launching searches in e-mails over being able to type two mentions after each other without spaces.

I'm not sure why that behaviour should change now.  Was that discussed in the issue perhaps?  I do agree that the negative lookahead is not necessary, since the word boundary match is a zero-character match anyway. `/\B@([\w-]*)$/` should have the exactly the same behaviour.

Or am I missing something?

(And for the other change, the addition of a capturing group for the username capture is a good idea. I guess you could argue to only match after the first character has been typed though, (So that would be `/\B@([\w-]+)$/`) but I guess that would also be a change from current behaviour)

Helcostr replied:
Oh dang! Where is the history?

Anyways, there was no bug report about this, but I removed it because I believe that it should work this way.

Plus it's not like each search sends a data request to Habitica...  so "phantom searches" in my book are fine. Especially when a user is typing fast. It's not like the logic of tagging is determined client side: it's done server side. And a no-space-tag is totally possible.

The user experience on my end seems fine... 🤔 

And no it's slightly different behavior... the search will include or not include the word boundary.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -128,38 +130,29 @@ export default {
+      )[caretPosition - (delCharFocusBool ? 0 : 1)];
+      if (
+        newTextParam.length === 0 // Delete all
+        || (charFocus === ' ' && !delCharFocusBool) // End Search
```
Is it enough to check on space characters for search end?  What about other types of whitespace characters?  Also I don't think the negative deletion check is necessary for spaces, at that point the search should have ended already anyway right?

```suggestion
        || /\s/.test(lastFocussed) // End Search
```

Helcostr replied:
Relic code I was mimicking, will implement this regex.

benkelaar commented on website/client/src/components/groups/questSidebarSection.vue:
```diff
@@ -149,7 +149,19 @@
+                      .toLocaleString(user.preferences.language, {
+                        minimumFractionDigits: 2,
+                        maximumFractionDigits: 2
+                      })
```
Would be good to externalize this functionality into a method.  Perhaps something like `localizeNumber(number, digits)`

Helcostr replied:
Okeydokey, I was wondering if I should do this.

benkelaar commented on website/client/src/components/groups/questSidebarSection.vue:
```diff
@@ -160,9 +172,15 @@
+                        maximumFractionDigits: 1
+                      })
+                  }} {{ $t('pendingDamageLabel') }}
+                </span> <!-- eslint-disable-line max-len -->
```
Line is no longer longer than max-len.
```suggestion
                </span>
```

Helcostr replied:
No idea how eslint works so I kept it in. Will remove.

benkelaar commented on website/client/src/components/ui/selectList.vue:
```diff
@@ -53,9 +55,19 @@
+    text-overflow: ellipsis;
+  }
+
   .select-list ::v-deep .selectListItem {
```
Here you can see the `::v-deep` selector being used, that did not work for you here?

Helcostr replied:
It didn't.

I can try again... maybe I might need to do a double v-deep 🤔...

Anyways, wrapping feels nicer XD. But I will try the v-deep method again to avoid too many DOM objects.

Edit: wow... v-deep worked 🤔 haha

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -170,11 +163,26 @@ export default {
     this.grabUserNames();
   },
   methods: {
+    solveSearchResults (searchTextWip) {
```
Is Wip for `Work in progress`?  Why not just `searchText`?

Helcostr replied:
cuz the text is being worked on at the moment (as opposed to the trailing end of the text that is dropped right b4 this parameter is passed into the method: that isn't being worked on).

tbh, I would not name it search. probably I would name it `textWip` since that's the client's focus currently.

Oh maybe I should name it `textFocus`?

benkelaar replied:
Yeah, might be a personal preference thing, I always try to put as little acronyms in my code as possible since they require more mental overhead to parse from the reader (and assume the reader knows about them in the first place).

Perhaps something along the lines of `textFilter` or `textQuery`?

Come to think of it `solve` is also a bit non-specific.  Perhaps `filterSearchResults` would be a better method name?

Helcostr replied:
Filter is weird. Query is more accurate, but still weird.

Search results is completely wrong.

I stuck with `textFocus` eventually.

benkelaar commented on website/client/src/components/chat/autoComplete.vue:
```diff
@@ -170,11 +163,26 @@ export default {
+      const lowerCaseSearch = this.currentSearch.toLowerCase();
+
+      return this.tmpSelections
+        .filter(option => { // eslint-disable-line arrow-body-style
+          return option.displayName.toLowerCase().indexOf(lowerCaseSearch) !== -1
+            || (option.username
+              && option.username.toLowerCase().indexOf(lowerCaseSearch) !== -1);
+        })
```
```suggestion
        .filter(option => option.displayName.toLowerCase().includes(lowerCaseSearch)
            || (option.username && option.username.toLowerCase().includes(lowerCaseSearch)))
```

Helcostr replied:
I think I maybe hitting the esline limit... gotta check

benkelaar reviewed (commented):
Awesome merge request!  Thanks a lot!

I added a couple suggestions for things I think would improve the code even more.

Helcostr committed:
Apply suggestions from code review

Build not complete at this point.

Co-authored-by: Bart Enkelaar <benkelaar@gmail.com>

Helcostr committed:
Apply suggestions from code review (part 2)
Build is ready for review at this point:
- Add filter for localizeNumber
- Use scss with v-deep properly (also pretty checkmark to "full-width" of checkmark)
- Clean various comments, and better variable names, and patches for compatability

Co-authored-by: Bart Enkelaar <benkelaar@gmail.com>

Helcostr commented:
@benkelaar am I suppose to re-request another review? There is a button to do so, but it's greyed out.

benkelaar commented:
> @benkelaar am I suppose to re-request another review? There is a button to do so, but it's greyed out.

Yeah, I can't "officially approve" the PR, since I'm just another contributor and as such my opinion doesn't matter to the merging process.

As far as I know, the person who was primarily responsible for handling these merges left for another job, so there's a bit of a backlog from the core team handling PRs.  I'm just trying to help out to get the PRs into the best shape possible so they don't have to spend too much time on them 😄.

As far as I know @SabreCat is the person now responsible for looking in to reviews, but I'm sure he'll get to it when he'll get to it 😃

Helcostr commented:
> Yeah, I can't "officially approve" the PR, since I'm just another contributor and as such my opinion doesn't matter to the merging process.
@benkelaar I was thinking of just requesting ur help with the review again... but I guess re-requesting is only available for those who actually have approving-the-PR-permissions?

benkelaar commented:
I guess it depends on how you "finalize" the review.  I didn't set the state to "request changes".  If I had I think you would have been able to re-request a review using that button.  But comments always work 😆

Helcostr commented:
This PR is ready for review.

Helcostr committed:
Re-add comment

comment was stripped during development, adding it back for notes

Helcostr committed:
Readd linebreak for pull consistency

Alys commented:
@Helcostr Thank you for your work on this! Could you please resubmit with one PR per issue? That will make it easier for us to test and review them, and it will also help if we ever need to refer back to how each issue was fixed (e.g., if a future code change makes the issue or a similar one re-appear, which happens sometimes).

Helcostr commented:
@Alys breakout done.
